base/loading.jl

```diff
@@ -2125,10 +2125,46 @@ macro __DIR__()
     return isempty(_dirname) ? pwd() : abspath(_dirname)
 end
 
+"""
+    Base.is_serializing_code()
+
+Returns whether julia is being run in a mode where code is being serialized
+to file. For instance during precompilation.
+
+For example, to run or precompile code only during package precompilation:
+```julia
+module FooBar
+
+foo(x::Int, y::Float64) = ...
+
+if Base.is_serializing_code()
+    foo(1, 1.0) # execute this method of foo to ensure it and any internal methods are precompiled
+
+    precompile(foo, (Int, Float64)) # precompile just this method of foo without executing it
+end
+end # module
+```
+
+Note that the first example here actually executes `foo(1, 1.0)` so can be slower, but is usually more robust
+to changes in the codebase than the latter approach of using `precompile` statements. The latter requires explicitly
+calling `precompile` on all methods, not just the entry method. i.e. `foo(1, 1.0)` will likely call other methods
```

Comment:
> The latter requires explicitly calling `precompile` on all methods, not just the entry method

Technically, only method dependencies that cannot be inferred. That is to say, suppose we have the following flattened call graph of `foo(::Int, ::Float64)`'s dependencies:

1. 47 MethodInstances are directly or indirectly called *inferrably*
2. 2 calls are made by runtime dispatch, and after inference of *those* entry points you collectively call 35 additional MethodInstances (all inferrably)

Then `precompile(foo, (Int, Float64))` would get the 47 inferrable MethodInstances, but would miss both the 2 runtime dispatch calls as well as their 35 dependencies. In contrast, `foo(1, 1.0)` would precompile them all. (Whether they get cached depends on whether the new entry points are owned by packages that are currently being serialized.) In other words, you could get the whole lot with 3 precompile directives, `foo(::Int, ::Float64)` and those two runtime-dispatched dependencies, because everything else can be deduced from them by type-inference.

Reply:
Ok. What about:
```suggestion
calling `precompile` on methods that cannot be inferred. i.e. `foo(1, 1.0)` will likely call other methods based on runtime dispatch
```

Reply:
Good, but instead of "will likely..." maybe "if `foo(1, 1.0)` calls some methods by runtime dispatch, the dispatched calls must be independently `precompile`d."

Reply:
FYI if you've used SnoopCompile's `inference_triggers`, these are the callee in each trigger and the base bar of the `@snoopi_deep` flames.

Reply:
updated. Thanks